### ISSUE TITLE
Email cannot be sent if CTF name has an accent

### ISSUE THREAD

**Issue by nymous** [CTFd/CTFd#1306]
**Environment**:

  - CTFd Version/Commit: 2.3.2
  - Operating System: Debian 10.3, Python 3.7.3
  - Web Browser and Version: Firefox Nightly, 76.0a1 (2020-04-06)

**What happened?**

After a fresh install of CTFd we set the competition name to "Inté CTF". We configure the email server to use our Postfix relay, and test it with the "forgotten password" feature.
The email never arrives and is bounced by relay servers because the `From` header is invalid.

The actual issue is because the `From` header has the form "CTF name <email address>" (see [here](https://github.com/CTFd/CTFd/blob/master/CTFd/utils/email/smtp.py#L25)). Once encoded in the `msg.as_string()` ([there](https://github.com/CTFd/CTFd/blob/master/CTFd/utils/email/smtp.py#L55)) it is entirely encoded in UTF-8 looking like this `=?utf-8?q?Int=C3=A9_CTF_=3Cctf-noreply=40example=2Ecom=3E?=`, which is invalid against RFC 2407 section 5 ("An 'encoded-word' MUST NOT appear in any portion of an 'addr-spec'"). A correct form would be `=?utf-8?q?Int=C3=A9?= CTF <ctf-noreply@example.com>`.

I have a patch ready but it uses `from email.message import EmailMessage` which is available in Python 3.6+. I'll be happy to open a PR if you confirm me that support for lower Python version can be dropped ^^

<details>
<summary>Patch for `CTFd/utils/email/smtp.py`</summary>

```diff
-from email.mime.text import MIMEText
+from email.message import EmailMessage


# around line 50
-        msg = MIMEText(text)
+        msg = EmailMessage()
+        msg.set_content(text)
         msg["Subject"] = subject
         msg["From"] = mailfrom_addr
         msg["To"] = addr
 
-        smtp.sendmail(msg["From"], [msg["To"]], msg.as_string())
+        smtp.send_message(msg)
```

</details>

**How to reproduce your issue**

- setup a new instance of CTFd
- configure the competition name with non-ascii characters
- configure the email server to use a "real" mail server (not Mailgun)
- try to send an email (with the "forgotten password" for example)
- see the mail server logs

<details>
<summary>
You can see the difference between both methods (`MIMEText` and `EmailMessage`) with the following snippet
</summary>

```python
from email.mime.text import MIMEText
msg = MIMEText("This is a message with accents éèçà")
msg["From"] = "René Côti <rene.coti@example.com>"
msg["To"] = "recipient@example.com"
msg.as_string()

from email.message import EmailMessage
msg = EmailMessage()
msg.set_content("This is a message with accents éèçà")
msg["From"] = "René Côti <rene.coti@example.com>"
msg["To"] = "recipient@example.com"
msg.as_string()
```

</details>

**Any associated stack traces or error logs**
```
# Sent to a gmail recipient
opendkim[839]: 278839FE69: can't parse From: header value ' =?utf-8?q?Int=C3=A9_CTF_=3Cctf-noreply=40example=2Ecom=3E?='
postfix/smtp[8076]: 278839FE69: to=<recipient@gmail.com>, relay=gmail-smtp-in.l.google.com[74.125.206.27]:25, delay=0.75, delays=0.31/0/0.23/0.21, dsn=5.7.1, status=bounced (host gmail-smtp-in.l.google.com[74.125.206.27] said: 550-5.7.1 [185.132.74.134      14] Messages missing a valid address in From: 550 5.7.1 header, or having no From: header, are not accepted. p64si4801636wmp.124 - gsmtp (in reply to end of DATA command))

# Sent to another recipient, with another relay server
postfix/smtpd[7858]: NOQUEUE: reject: RCPT from mailcube2.domain.fr[193.51.52.6]: 550 5.1.1 <ctf-noreply@example.com>: Recipient address rejected: User unknown
 in virtual mailbox table; from=<> to=<ctf-noreply@example.com> proto=ESMTP helo=<mailcube2.domain.fr>
```

**Comment by ColdHeat** [comment]
Hmmmm I haven't dug into this yet but I just tried to reproduce it using a hosted CTFd instance which uses SMTP and I wasn't able to reproduce. Messages with UTF-8 went through fine. 

Can you replicate this issue with any other SMTP provider? Maybe try gmail? Or if there's some kind of email validator?

I do believe that this is probably an issue but we can't drop Python 2 support yet. Is there a Python 2 compatible fix that would work here?

**Comment by nymous** [comment]
Just to give you an update, I did test using my personal Gmail account as the SMTP server, and it did work correctly. Maybe Gmail is more lax on the headers parsing, maybe Postfix is plain buggy, I don't know.

What I did observe though is that the `MIMEText.as_string()` behavior is different on Python 2: it only encodes the "name" part of the From header, giving something like `Int\xc3\xa9 CTF <ctf-noreply@example.com>`. I did not test to see if it was accepted by our Postfix server.

If it is valid, a simple workaround would be to run different code on Python 2 and 3, like this:
```python
import sys
if sys.version_info.major < 3:
from email.mime.text import MIMEText
    from email.mime.text import MIMEText
else:
    from email.message import EmailMessage


    try:
        smtp = get_smtp(**data)
        if sys.version_info.major < 3:
            msg = MIMEText(text)
        else:
            msg = EmailMessage()
            msg.set_content(text)
        msg["Subject"] = subject
        msg["From"] = mailfrom_addr
        msg["To"] = addr

        if sys.version_info.major < 3:
            smtp.sendmail(msg["From"], [msg["To"]], msg.as_string())
        else:
            smtp.send_message(msg)
        smtp.quit()
        return True, "Email sent"
```

Not sure if this is clean enough, or if it's better to wait for the 3.0 (which will drop Python 2 support I guess?)

As a side note, during my investigation I got lost in the stdlib code ^^', and the code for header encoding is a lot more complex on an EmailMessage in Python 3, it has specific parsers for different headers (From is not the same as Subject for example). I tried to reproduce a similar behavior using `email.header` and encoding part of the address manually, but without result yet.

**Comment by ColdHeat** [comment]
Thanks for doing the investigation on this! 

I think as long as there's a fallback for Python 2 this can go in earlier than v3. It would be nice to understand what the underlying issue is but it's good to know that it's fixed with Python 3. 

As I mentioned before, I think I've seen errors in this SMTP email sending area so I'm willing to push this up and then remove the legacy Python 2 code once v3 hits. The only concern here is that it's a little more difficult to test code that calls two different sending functions. If `EmailMessage` resolves the other issues I've seen I'll definitely patch this before v3. 

As an aside, I would probably argue that this is something to do with Postfix. I imagine gmail sees weird clients and knows to be a little more flexible.

**Comment by ColdHeat** [comment]
Perhaps but the docs suggest that send_message has a few other pieces of code in it that wrap sendmail. I imagine we'll just have to do both.

**Comment by ColdHeat** [comment]
@nymous Looking at the test output from switching to EmailMessage, I suspect maybe MIMEText needs to be set to use utf-8 as the email charset. 

Do you think you can try `msg = MIMEText(text, _charset="utf-8")` to see if that would work on your Postfix server? 

Test output:

```
E   AssertionError: Expected call: sendmail('CTFd <noreply@ctfd.io>', ['user@user.com'], 'Content-Type: text/plain; charset="us-ascii"\nMIME-Version: 1.0\nContent-Transfer-Encoding: 7bit\nSubject: Message from CTFd\nFrom: CTFd <noreply@ctfd.io>\nTo: user@user.com\n\nthis is a test')
E   Actual call: sendmail('CTFd <noreply@ctfd.io>', ['user@user.com'], 'Content-Type: text/plain; charset="utf-8"\nContent-Transfer-Encoding: 7bit\nMIME-Version: 1.0\nSubject: Message from CTFd\nFrom: CTFd <noreply@ctfd.io>\nTo: user@user.com\n\nthis is a test\n')
```

**Comment by ColdHeat** [comment]
Hmm setting the charset actually may work as it appears it's the only difference between sendmail and smtp_message:

Python 3 with smtp_message
```
tests/utils/test_email.py Content-Type: text/plain; charset="utf-8"
Content-Transfer-Encoding: 7bit
MIME-Version: 1.0
Subject: Message from CTFd
From: Inté CTF <CTFd>
To: user@user.com

this is a test
```

Python 2 with sendmail
```
tests/utils/test_email.py From nobody Fri Apr 17 15:25:49 2020
Content-Type: text/plain; charset="us-ascii"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: Message from CTFd
From: Inté CTF <CTFd>
To: user@user.com

this is a test
```

**Comment by ColdHeat** [comment]
I was able to test this out and EmailMessage resolves a lot of problems. Setting the charset doesn't really help that much and doesn't resolve this problem either. 

With that in mind, this is one of those warts that's going to have to have a version check and then fixed in v3.

**Comment by ColdHeat** [comment]
Related to  #820. The version checking will need to be removed in v3.

**Comment by nymous** [comment]
Thank you for your work and for this fix! I'll see to test this soon!

**Comment by ColdHeat** [comment]
This should be closed with #1330 in the 2.4.0-dev branch.
